bindings/ocaml/transforms/ipo/ipo_ocaml.c

```diff
@@ -91,6 +96,34 @@ CAMLprim value llvm_add_internalize(LLVMPassManagerRef PM, value AllButMain) {
   return Val_unit;
 }
 
+/* string -> bool */
+static value *predicate_f = NULL;
+
+LLVMBool MustPreserveCallBack(LLVMValueRef Val, void* Ctx) {
+  CAMLparam0();
+  CAMLlocal1(LLVMValName);
+  const char *llvmValName;
+
+  assert(predicate_f != NULL
+         && "llvm_add_internalize_predicate must be called with \
+             LLVMInternalizePredicateCallback symbol set");
+
+  llvmValName = LLVMGetValueName(Val);
+  LLVMValName = caml_copy_string(llvmValName);
+  LLVMBool ret = Bool_val(caml_callback(*predicate_f, LLVMValName));
+
+  CAMLreturnT(LLVMBool, ret);
+}
+
+/* [`Module] Llvm.PassManager.t -> unit */
+CAMLprim value llvm_add_internalize_predicate(LLVMPassManagerRef PM) {
+  if (predicate_f == NULL)
```

Comment:
If this is ever called with two different predicates, only the first one will be used. I think that predicate_f should be cleared somehow, or always set, or something.

Reply:
That is a very good point. I think always setting it might be the best option, this is not hot bit of code.